File: BST.cpp
#include <iostream>
#include <iomanip>
#include <cstring>
#include <cmath>
#include <cstdio>
#include <algorithm>
#include <vector>
#include <climits>
#include <stack>
#include <queue>
#include <sstream>
#include <map>
using namespace std;

class Node
{
	public:
		Node* left;
		Node* right;
		int data;
	public:
		Node(int val)
		{
			data = val;
		}
};
class BST
{
	private:
		Node* root;
	public:
		void insert(int val)
		{
			Node* newNode = new Node(val);
			Node* itr = root;
			if(root == NULL)
				root = newNode;
			else
			{
				while(true)
				{
					if(val > itr->data)
					{
						if(itr->right == NULL)
						{
							itr->right = newNode;
							return ;
						}
						itr = itr->right;

					}
					else if(val < itr->data)
					{
						if(itr->left == NULL)
						{
							itr->left = newNode;
							return ;
						}
						itr = itr->left;
					}
				}
			}
		}
		Node* getRoot()
		{
			return root;
		}
		void findMax()
		{
			Node* itr = root;
			while(itr->right != NULL)
				itr = itr->right;
			cout << itr->data << endl;
		}
		void findMin()
		{
			Node* itr = root;
			while(itr->left != NULL)
				itr = itr->left;
			cout << itr->data << endl;
		}
		void inorder(Node* itr)
		{
			if(itr == NULL)
				return;
			if(itr->left != NULL)
				inorder(itr->left);
			cout << itr->data << endl;
			if(itr->right != NULL)
				inorder(itr->right);
		}
};
int main()
{
	BST* bst = new BST();
	bst->insert(15);
	bst->insert(4);
	bst->insert(9);
	bst->insert(55);
	bst->insert(22);
	cout << "test" << endl;
	bst->findMax();
	bst->findMin();
}

File: BST.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */

package bst;

/**
 *
 * @author Science
 */
public class Main {
    public static void main(String[]args)
    {
        BinarySearchTree bst = new BinarySearchTree();
        bst.insert(6);
        bst.insert(4);
        bst.insert(10);
        bst.insert(3);
        bst.insert(5);
        bst.insert(9);
        bst.insert(11);
        bst.inorder(bst.root);
    }

}
class BinarySearchTree
{


    class Node
    {
        int data;
        Node left,right;
        Node(int _data)
        {
            data = _data;
        }
    }

    Node root;
    public void insert(int data)
    {
        Node newNode = new Node(data);
        Node itr = root;
        if(root == null)
        {
            root = newNode;
        }
        else
        {
            while(true)
            {
                if(data > itr.data)
                {
                    if(itr.right == null)
                    {
                        itr.right = newNode;
                        return;
                    }
                    itr = itr.right;
                }
                else
                {
                    if(itr.left == null)
                    {
                        itr.left = newNode;
                        return;
                    }
                    itr = itr.left;
                }
            }
        }
    }
    public void deleteData(int data)
    {
        root = deleteRec(root,data);
    }
    public Node deleteRec(Node root,int data)
    {
        if(root == null)
            return root;
        if(data < root.data)
        {
            root.left = deleteRec(root.left,data);
        }
        else if(data > root.data)
        {
            root.right = deleteRec(root.right,data);
        }
        else
        {
            if(root.left == null)
                return root.right;
            else if(root.right == null)
                return root.left;

            root.data = minValue(root.right);
            root.right = deleteRec(root.right,root.data);
        }
        return root;
    }
    public int minValue(Node root)
    {
        int _min = root.data;
        while(root.left != null)
        {
            _min = root.left.data;
            root = root.left;
        }
        return _min;
    }
    public void inorder(Node node)
    {
        if(node == null)
            return ;
        inorder(node.left);
        System.out.println(node.data);
        inorder(node.right);
    }
}
